svm/src/transaction_processor.rs

```diff
@@ -1292,6 +1299,165 @@ mod tests {
+                    let next_program_address = PROGRAM_ADDRESSES[next_program_index];
+                    let accounts = (next_program_index..PROGRAM_ADDRESSES.len())
+                        .map(|index| AccountMeta::new_readonly(PROGRAM_ADDRESSES[index], false))
+                        .collect_vec();
+                    last_result = invoke_context.native_invoke(
+                        Instruction::new_with_bytes(
+                            next_program_address,
+                            &[
+                                index_of_program_that_must_throw_exception as u8,
+                                next_program_index as u8,
+                            ],
+                            accounts,
+                        )
+                        .into(),
+                        &[],
+                    );
+                    if last_result.is_err() {
+                        return last_result;
+                    }
+                }
+                return last_result;
+            }
+
+            Ok(())
+        });
+
+        // =======================================================
+        // BEGIN: Create a transaction that calls the mock program
+        // =======================================================
+        let base_program_index = 0;
+        let accounts = (0..PROGRAM_ADDRESSES.len())
+            .map(|index| AccountMeta::new_readonly(PROGRAM_ADDRESSES[index], false))
+            .collect_vec();
+        let message: Message = Message::new(
+            &[Instruction::new_with_bytes(
+                PROGRAM_ADDRESSES[base_program_index],
+                &[
+                    index_of_program_that_should_throw_exception,
+                    base_program_index as u8, /* index of program being called */
+                ],
+                accounts,
+            )],
+            None,
+        );
+        let sanitized_message = new_unchecked_sanitized_message(message);
+        let mut program_cache_for_tx_batch = ProgramCacheForTxBatch::default();
+        for index in 0..PROGRAM_ADDRESSES.len() {
+            program_cache_for_tx_batch.replenish(
+                PROGRAM_ADDRESSES[index],
+                Arc::new(ProgramCacheEntry::new_builtin(0, 0, MockBuiltin::vm)),
+            );
+        }
+        let batch_processor = TransactionBatchProcessor::<TestForkGraph>::default();
+        let sanitized_transaction = SanitizedTransaction::new_for_tests(
+            sanitized_message,
+            vec![Signature::new_unique()],
+            false,
+        );
+        let mut mock_program_account = AccountSharedData::new(1, 0, &native_loader::id());
+        mock_program_account.set_executable(true);
+        let loaded_transaction = LoadedTransaction {
+            accounts: (0..PROGRAM_ADDRESSES.len())
+                .map(|index| (PROGRAM_ADDRESSES[index], mock_program_account.clone()))
+                .collect_vec(),
```

Comment:
This is how we know that the program accounts are in the order we expect them to be from the perspective of the transaction.